### ISSUE TITLE
Questions about ensembler.py

### ISSUE THREAD

**Issue by Superhzf** [sadz2201/Kaggle_Allstate#2]
Sorry for having so many questions. I cannot understand what is going on here. 

My questions are

1. What does the function `myScaler` do especially the parameters width and start and How did you determine them? It seems that this function does not use any oof values but just the order of predictions. 

2. Why did you just input A, B and E into myScaler, how about C, D and F?

**Comment by sadz2201** [comment]
Ok I'll try to answer this as best as I can.
Question 1:-
I analyzed the distribution of my test & oof predictions, & compared it with the distribution of the actual loss value from the train set. I found that the predicted min & 25 percentile are consistently larger than the actual, & the predicted 75 percentile & max are consistently smaller than actual. Thus I had to come up with a correction function which will scale down the low values (bcoz they aren't low enough), & scale up the high values (bcoz they aren't high enough). In essence, the scaling factor (or multiplier) for any value depends on it's percentile. 

As an example, if you use width=0.1, start =0.95: - 
new_min=old_min * 0.95. 
new_25perc= old_25perc * 0.975
median -- remains as is. 
new_max = old_max * 1.05. All values in between are scaled by a proportionate multiplier.

The advantage of passing both width & start as parameters is you can try asymmetric transforms. (eg width 0.08, start 0.95. in this case new_median = old_median * 0.99)

Obviously, start & width are parameters which you have to optimize. I only uploaded my cleaned up code, but you could do something like this: 
```
A=pd.read_csv('output/my_xgb1_oof_tr.csv')
best=mean_absolute_error(actual.loss,A.loss)
width=0,start=1
start_list=[list_of_start_values_to_try]
width_list=[list_of_width_values_to_try]
for i,j in width_list,start_list:
      A_t=myScaler(A.copy(),i,j)
      score=mean_absolute_error(actual.loss,A_t.loss)
      if score<best:
              best=score
              width=i
              start=j

#now use these start & width values to transform my_xgb1 test predictions
#repeat similarly for all other base models
```

Question 2:
This scaling hack worked really well for XGB & LightGBM models, but made things worse for Keras models. (I did burn a few submissions testing it for Keras. Both CV & LB scores were worse than without the transform). Hence I'm using it for only models A,B & E.
